File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/cassandra/CassandraNativeTypes.java
package ru.tinkoff.kora.database.annotation.processor.cassandra;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.TypeName;

import javax.annotation.Nullable;
import java.math.BigDecimal;
import java.nio.ByteBuffer;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.List;
import java.util.Objects;

public class CassandraNativeTypes {
    private static final List<CassandraNativeType> nativeTypes;

    static {
        var booleanPrimitive = CassandraNativeType.of(
            TypeName.BOOLEAN,
            (rsName, i) -> CodeBlock.of("$L.getBoolean($L)", rsName, i),
            (stmt, var, i) -> CodeBlock.of("$L.setBoolean($L, $L)", stmt, i, var)
        );
        var booleanBoxed = booleanPrimitive.boxed();
        var intPrimitive = CassandraNativeType.of(
            TypeName.INT,
            (rsName, i) -> CodeBlock.of("$L.getInt($L)", rsName, i),
            (stmt, var, i) -> CodeBlock.of("$L.setInt($L, $L)", stmt, i, var)
        );
        var intBoxed = intPrimitive.boxed();
        var longPrimitive = CassandraNativeType.of(
            TypeName.LONG,
            (rsName, i) -> CodeBlock.of("$L.getLong($L)", rsName, i),
            (stmt, var, i) -> CodeBlock.of("$L.setLong($L, $L)", stmt, i, var)
        );
        var longBoxed = longPrimitive.boxed();
        var doublePrimitive = CassandraNativeType.of(
            TypeName.DOUBLE,
            (rsName, i) -> CodeBlock.of("$L.getDouble($L)", rsName, i),
            (stmt, var, i) -> CodeBlock.of("$L.setDouble($L, $L)", stmt, i, var)
        );
        var doubleBoxed = doublePrimitive.boxed();
        var string = CassandraNativeType.of(
            ClassName.get(String.class),
            (rsName, i) -> CodeBlock.of("$L.getString($L)", rsName, i),
            (stmt, var, i) -> CodeBlock.of("$L.setString($L, $L)", stmt, i, var)
        );
        var bigDecimal = CassandraNativeType.of(
            ClassName.get(BigDecimal.class),
            (rsName, i) -> CodeBlock.of("$L.getBigDecimal($L)", rsName, i),
            (stmt, var, i) -> CodeBlock.of("$L.setBigDecimal($L, $L)", stmt, i, var)
        );
        var byteBuffer = CassandraNativeType.of(
            ClassName.get(ByteBuffer.class),
            (rsName, i) -> CodeBlock.of("$L.getByteBuffer($L)", rsName, i),
            (stmt, var, i) -> CodeBlock.of("$L.setByteBuffer($L, $L)", stmt, i, var)
        );
        var localDateTime = CassandraNativeType.of(
            TypeName.get(LocalDateTime.class),
            (rsName, i) -> CodeBlock.of("$L.get($L, $T.class)", rsName, i, LocalDateTime.class),
            (stmt, var, i) -> CodeBlock.of("$L.set($L, $L, $T.class)", stmt, i, var, LocalDateTime.class)
        );
        var localDate = CassandraNativeType.of(
            TypeName.get(LocalDate.class),
            (rsName, i) -> CodeBlock.of("$L.getLocalDate($L)", rsName, i),
            (stmt, var, i) -> CodeBlock.of("$L.setLocalDate($L, $L)", stmt, i, var)
        );
        var instant = CassandraNativeType.of(
            TypeName.get(Instant.class),
            (rsName, i) -> CodeBlock.of("$L.getInstant($L)", rsName, i),
            (stmt, var, i) -> CodeBlock.of("$L.setInstant($L, $L)", stmt, i, var)
        );
        var localTime = CassandraNativeType.of(
            TypeName.get(LocalTime.class),
            (rsName, i) -> CodeBlock.of("$L.getLocalTime($L)", rsName, i),
            (stmt, var, i) -> CodeBlock.of("$L.setLocalTime($L, $L)", stmt, i, var)
        );

        nativeTypes = List.of(
            booleanPrimitive,
            booleanBoxed,
            intPrimitive,
            intBoxed,
            longPrimitive,
            longBoxed,
            doublePrimitive,
            doubleBoxed,
            string,
            bigDecimal,
            byteBuffer,
            localDateTime,
            localDate,
            instant,
            localTime
        );
    }

    @Nullable
    public static CassandraNativeType findNativeType(TypeName typeName) {
        for (var nativeParameterType : nativeTypes) {
            if (Objects.equals(nativeParameterType.type(), typeName)) {
                return nativeParameterType;
            }
        }
        return null;
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/cassandra/CassandraNativeTypes.kt
package ru.tinkoff.kora.database.symbol.processor.cassandra

import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.asTypeName
import java.math.BigDecimal
import java.nio.ByteBuffer
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime

object CassandraNativeTypes {
    private val nativeTypes = listOf(
        CassandraNativeType.of(
            Boolean::class.asTypeName(),
            { rsName, i -> CodeBlock.of("%N.getBoolean(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setBoolean(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            Boolean::class.asTypeName().copy(true),
            { rsName, i -> CodeBlock.of("%N.getBoolean(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setBoolean(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            Int::class.asTypeName(),
            { rsName, i -> CodeBlock.of("%N.getInt(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setInt(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            Int::class.asTypeName().copy(true),
            { rsName, i -> CodeBlock.of("%N.getInt(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setInt(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            Long::class.asTypeName(),
            { rsName, i -> CodeBlock.of("%N.getLong(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setLong(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            Long::class.asTypeName().copy(true),
            { rsName, i -> CodeBlock.of("%N.getLong(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setLong(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            Double::class.asTypeName(),
            { rsName, i -> CodeBlock.of("%N.getDouble(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setDouble(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            Double::class.asTypeName().copy(true),
            { rsName, i -> CodeBlock.of("%N.getDouble(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setDouble(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            String::class.asTypeName(),
            { rsName, i -> CodeBlock.of("%N.getString(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setString(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            String::class.asTypeName().copy(true),
            { rsName, i -> CodeBlock.of("%N.getString(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setString(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            BigDecimal::class.asTypeName(),
            { rsName, i -> CodeBlock.of("%N.getBigDecimal(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setBigDecimal(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            BigDecimal::class.asTypeName().copy(true),
            { rsName, i -> CodeBlock.of("%N.getBigDecimal(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setBigDecimal(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            ByteBuffer::class.asTypeName(),
            { rsName, i -> CodeBlock.of("%N.getByteBuffer(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setByteBuffer(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            ByteBuffer::class.asTypeName().copy(true),
            { rsName, i -> CodeBlock.of("%N.getByteBuffer(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setByteBuffer(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            LocalDate::class.asTypeName(),
            { rsName, i -> CodeBlock.of("%N.getLocalDate(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setLocalDate(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            LocalDate::class.asTypeName().copy(true),
            { rsName, i -> CodeBlock.of("%N.getLocalDate(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setLocalDate(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            LocalDateTime::class.asTypeName(),
            { rsName, i -> CodeBlock.of("%N.get(%L, %T::class.java)", rsName, i, LocalDateTime::class) },
            { stmt, variableName, idx -> CodeBlock.of("%N.set(%L, %L, %T::class.java)", stmt, idx, variableName, LocalDateTime::class) },
        ),
        CassandraNativeType.of(
            LocalDateTime::class.asTypeName().copy(true),
            { rsName, i -> CodeBlock.of("%N.get(%L, %T::class.java)", rsName, i, LocalDateTime::class) },
            { stmt, variableName, idx -> CodeBlock.of("%N.set(%L, %L, %T::class.java)", stmt, idx, variableName, LocalDateTime::class) },
        ),
        CassandraNativeType.of(
            Instant::class.asTypeName(),
            { rsName, i -> CodeBlock.of("%N.getInstant(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setInstant(%L, %L)", stmt, idx, variableName) },
        ),
        CassandraNativeType.of(
            Instant::class.asTypeName().copy(true),
            { rsName, i -> CodeBlock.of("%N.getInstant(%L)", rsName, i) },
            { stmt, variableName, idx -> CodeBlock.of("%N.setInstant(%L, %L)", stmt, idx, variableName) },
        ),
    )

    fun findNativeType(typeName: TypeName): CassandraNativeType? {
        for (nativeParameterType in nativeTypes) {
            if (nativeParameterType.type() == typeName) {
                return nativeParameterType
            }
        }
        return null
    }
}
